enterprise/internal/a8n/webhooks.go

```diff
@@ -12,18 +12,89 @@ import (
 	gh "github.com/google/go-github/v28/github"
 	"github.com/sourcegraph/sourcegraph/cmd/repo-updater/repos"
 	"github.com/sourcegraph/sourcegraph/internal/a8n"
+	// "github.com/sourcegraph/sourcegraph/internal/extsvc/bitbucketserver"
+	"github.com/sourcegraph/sourcegraph/internal/extsvc/bitbucketserver"
 	"github.com/sourcegraph/sourcegraph/internal/extsvc/github"
 	"github.com/sourcegraph/sourcegraph/schema"
 	"gopkg.in/inconshreveable/log15.v2"
 )
 
+type Webhook struct {
+	Store *Store
+	Repos repos.Store
+	Now   func() time.Time
+
+	Service string
+}
+
+func (h Webhook) upsertChangesetEvent(
+	ctx context.Context,
+	pr int64,
+	ev interface{ Key() string },
+) (err error) {
+	var tx *Store
+	if tx, err = h.Store.Transact(ctx); err != nil {
+		return err
+	}
+
+	defer tx.Done(&err)
+
+	cs, err := tx.GetChangeset(ctx, GetChangesetOpts{
+		ExternalID:          strconv.FormatInt(pr, 10),
+		ExternalServiceType: h.Service,
+	})
+	if err != nil {
+		if err == ErrNoResults {
+			err = nil // Nothing to do
+		}
+		return err
+	}
+
+	now := h.Now()
+	event := &a8n.ChangesetEvent{
+		ChangesetID: cs.ID,
+		Kind:        a8n.ChangesetEventKindFor(ev),
+		Key:         ev.Key(),
+		CreatedAt:   now,
+		UpdatedAt:   now,
+		Metadata:    ev,
+	}
+
+	existing, err := tx.GetChangesetEvent(ctx, GetChangesetEventOpts{
+		ChangesetID: cs.ID,
+		Kind:        event.Kind,
+		Key:         event.Key,
+	})
+
+	if err != nil && err != ErrNoResults {
+		return err
+	}
+
+	if existing != nil {
+		existing.Update(event)
+		event = existing
+	}
+
+	return h.Store.UpsertChangesetEvents(ctx, event)
```

Comment:
Shouldn't this use the transaction? Otherwise this code would be called _before_ `tx.Done()` is called, right?

```suggestion
	return tx.UpsertChangesetEvents(ctx, event)
```

Reply:
It should indeed. If this was already like this, then it's a bug.

Reply:
Wait... Is this the first time I spotted something that @unknwon didn't see before?

![](https://media.giphy.com/media/JRDq6qnIaqbMyIsqwd/giphy.gif)